### ISSUE TITLE
Check for nulls in Native.java

### ISSUE THREAD

**Issue by cheshire** [Z3Prover/z3#673]
Giving null (0 in Java-land) to Z3 results in segfaults.
Segfaults are very hard to debug as they don't provide a traceback.

Throwing exceptions on 0-inputs in Native.java would make debugging much easier. 
Is Native.java generated by a python script?
I could try to do that, but I'm not sure which parameters accept nulls and which don't.

**Comment by NikolajBjorner** [comment]
Yes, Native.java is created by a python script  in the scripts/ directory.
The code above, though, seems currently to own the parameter validation story 
at the level of managed pointers.

**Comment by cheshire** [comment]
@NikolajBjorner what do you mean by "the code above", the Java wrapping exposed to the end user? Would you mind a pull request shifting the validation down to the JNI code?

**Comment by NikolajBjorner** [comment]
"code above" = "code above the C API", which includes Java code. If you pass a null pointer to a Java method, it may throw a Java null pointer exception well before Z3's C API gets called. So the parameter validation should happen at the Java level. Some parameter validation happens, but it is unlikely complete. A more accurate description is that it is best-effort.

For one example, the andThen tactic just assumes the tactic arrays are non-null tactics. It accesses each element in the array. If the elements are null, then Java throws.
For a different example, Z3Object.arrayToNative checks for null pointers and assumes it is legal to pass null asts in this case. We presume the native Z3 code handles such cases gracefully by throwing or accepting such input when it makes sense.

In summary, I think pull requests to help improve consistency and resilience based on cases where users found it instrumental would be welcome.

**Comment by wintersteiger** [comment]
The C layer does verify arguments in many cases (but possibly not all), i.e., at the Java level there will be Z3Exception with the message "invalid argument". 

I suggest we should make sure that all arguments are at least validated on the C level to avoid segfaults in all APIs. For now I think we can live with Z3Exceptions instead of API-specific exceptions like NullPointerException.
